File: packages/core/src/temporal.ts
import type { TemporalRecord } from "./types";

export function isIsoDateTime(value: string): boolean {
  if (typeof value !== "string" || value.length < 10) return false;
  const parsed = Date.parse(value);
  return Number.isFinite(parsed) && /T.*(?:Z|[+-]\d{2}:?\d{2})$/.test(value);
}

export function toEpoch(value: string): number {
  const parsed = Date.parse(value);
  if (!Number.isFinite(parsed)) {
    throw new TypeError(`Invalid date-time: ${value}`);
  }
  return parsed;
}

export function isKnownAt(record: TemporalRecord, knownAt: string): boolean {
  const point = toEpoch(knownAt);
  if (toEpoch(record.observedAt) > point) return false;
  return record.supersededAt ? toEpoch(record.supersededAt) > point : true;
}

export function isValidAt(record: TemporalRecord, validAt: string): boolean {
  const point = toEpoch(validAt);
  if (toEpoch(record.validFrom) > point) return false;
  return record.validTo ? toEpoch(record.validTo) > point : true;
}

export function isVisibleAt(
  record: TemporalRecord,
  validAt: string,
  knownAt: string,
): boolean {
  return isKnownAt(record, knownAt) && isValidAt(record, validAt);
}

File: src/cascadelens/temporal.py
"""Bitemporal parsing and visibility helpers."""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any


def parse_datetime(value: str) -> datetime:
    if not isinstance(value, str) or not value:
        raise ValueError("Expected an ISO-8601 date-time with timezone")
    normalized = value[:-1] + "+00:00" if value.endswith("Z") else value
    parsed = datetime.fromisoformat(normalized)
    if parsed.tzinfo is None or parsed.utcoffset() is None:
        raise ValueError("Expected an ISO-8601 date-time with timezone")
    return parsed.astimezone(UTC)


def is_datetime(value: Any) -> bool:
    try:
        parse_datetime(value)
    except (TypeError, ValueError):
        return False
    return True


def isoformat(value: datetime) -> str:
    return value.astimezone(UTC).isoformat(timespec="seconds").replace("+00:00", "Z")


def is_known_at(record: dict[str, Any], known_at: str) -> bool:
    point = parse_datetime(known_at)
    observed = parse_datetime(record["observedAt"])
    superseded = record.get("supersededAt")
    return observed <= point and (superseded is None or point < parse_datetime(superseded))


def is_valid_at(record: dict[str, Any], valid_at: str) -> bool:
    point = parse_datetime(valid_at)
    start = parse_datetime(record["validFrom"])
    end = record.get("validTo")
    return start <= point and (end is None or point < parse_datetime(end))


def is_visible_at(record: dict[str, Any], valid_at: str, known_at: str) -> bool:
    return is_known_at(record, known_at) and is_valid_at(record, valid_at)
